🚀 3단계 - 지하철 구간 관리

hhhhhwi opened the pull request:
안녕하세요 혜영님..!

3단계 PR이 좀 늦었습니다 ㅠ
늦은 만큼 좀 더 좋은 코드를 작성하고 싶었는데 제 맘대로 잘 안되네요 ㅠ

욕심을 버리고 일단 피드백을 받는게 좋을 것 같아 리뷰 요청 드립니다!

hhhhhwi committed:
test: 지하철 구간 등록 인수 테스트 구현

hhhhhwi committed:
refactor: Line 객체의 상행 종점역, 하행 종점역 변수명 변경
- upStationId -> startStationId
- downStationId -> endStationId

hhhhhwi committed:
feat: 지하철 구간 객체 및 등록 기능 구현

hhhhhwi committed:
test: 지하철 구간 등록 실패 테스트 구현

hhhhhwi committed:
refactor: dto 객체로 테스트 하도록 수정

hhhhhwi committed:
test: 지하철 구간 삭제 인수 테스트 구현

hhhhhwi committed:
test: 지하철 구간 삭제 실패 테스트 구현

hhhhhwi committed:
refactor: 지하철 노선 등록 및 삭제 시 해당 구간도 등록 및 삭제 되도록 수정

hhhhhwi committed:
refactor: 구간 인수 테스트 시, sql로 노선 등록을 직접 api 호출하도록 수정

hhhhhwi committed:
refactor: delete 변수 별 요청에 따라 오버로딩

hhhhhwi committed:
feat: 지하철 구간 삭제 기능 구현

hhhhhwi committed:
refactor: 중복 메소드 리팩토링

hhhhhwi committed:
refactor: Section 엔티티에 대한 Unique 컬럼 설정 추가

hhhhhwi commented on src/main/java/subway/service/LineService.java:
```diff
@@ -52,11 +59,69 @@ public void updateLine(Long id, LineRequest lineRequest) {
+		Line line = findLindById(id);
+
+		if(!line.getEndStationId().equals(sectionRequest.getUpStationId())) {
+			throw new IllegalArgumentException("노선의 하행 종점역과 구간의 상행역은 같아야 합니다.");
```
지금은 서비스 레이어에서 Exception을 던지고 있는데,

사실 해당 부분은 입력값에 대한 validation 이 아닐까? 라는 고민과
해당 Exception을 단순히 Internal Server Error (500) 로 보여주는게 좀 아쉽습니다 ㅠ
어떻게 처리하면 좋을까요..?

hiblue02 replied:
1, `노선의 하행 종점역과 구간의 상행역은 같아야 한다`는 건 Line을 조회해야 검증할 수 있는데요. 단순한 길이/null 체크와 같은 검증이라기 보단 비즈니스 요건에 가깝다고 생각해요. 

비즈니스 요건이라면! 서비스에 두는 게 좋을까요? entity에 두는게 좋을까요?  

2. 단순히 500에러만 내기 아쉽다면 ControllerAdvice를 이용해 exception 메세지를 보여주는 건 어떨까요? 

https://tecoble.techcourse.co.kr/post/2023-05-03-ExceptionHandler-ControllerAdvice/

hhhhhwi commented on src/test/java/subway/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,150 @@
+	@DisplayName("지하철 노선에 등록되어 있는 역을 등록하면 실패한다.")
+	@DirtiesContext
+	@Test
+	void createSectionWithRegisteredStationThenFailTest() {
```
이번엔 실패 시나리오에 대한 테스트 메소드까지 같이 작성하고 구현하는 방식으로 진행했는데요..!
메소드 명.. 을 더 명확하게 작성하는 컨벤션 같은 게 있을까요..?
그리고 HttpStatus 코드값 검증 외에 더 검증해야 될 게 있을까요?!

hiblue02 replied:
저는 에러 메세지가 잘 나왔는지 검증합니다.  :)

hiblue02 commented on src/main/java/subway/service/LineService.java:
```diff
@@ -52,11 +59,69 @@ public void updateLine(Long id, LineRequest lineRequest) {
+			if (sectionRequest.getDownStationId().equals(x.getUpStationId()) || sectionRequest.getDownStationId().equals(x.getDownStationId())) {
+				throw new IllegalArgumentException("해당 노선에 " + sectionRequest.getDownStationId() + "역이 이미 존재합니다.");
+			}
+		});
```
1. if 안의 조건 문장이 길어졌네요. :sob:
section의 upStationid, downStationId를 꺼내서 밖에서 비교하고 있어요. 이러면 객체의 캡슐화가 깨지게 됩니다.  이미 있는 역인지, section에게 물어보는 건 어떨까요?

2. Line에 Section을 추가해주면 sectionRepository로 직접 조회하는 작업을 하지 않아도 됩니다. :sob: 

3. 2주차로 가면, Line이 갖고 있는 section들을 조회 후 검증하는 부분이 많아집니다. section을 일급컬렉션으로 구성해보는 건 어떨까요?

hiblue02 commented on src/main/java/subway/service/LineService.java:
```diff
@@ -52,11 +59,69 @@ public void updateLine(Long id, LineRequest lineRequest) {
+			throw new IllegalArgumentException("상행 종점역과 하행 종점역만 있는 노선입니다.");
+		}
+
+		sectionRepository.delete(section);
+
+		line.removeSection(section.getUpStationId(), section.getDistance());
+		lineRepository.save(line);
```
1. Line과 Section을 직접 참조로 연결했기 때문에 Line에서 removeSection만 호출해도 section이 삭제되지 않나요? (+JPA cascade 옵션)

~~~java
    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
    @JoinColumn(name = "line_id")
    private Line line;
~~~

2. 객체가 만들어거나 수정되면 save를 하지 않아도 jpa 영속성 컴포넌트가 알아서 db에 반영해줍니다. 코드를 깔끔하게 바꾸고 싶다면 save 부분울 지워보는 방법도 있습니다.

hiblue02 commented on src/main/java/subway/service/LineService.java:
```diff
@@ -52,11 +59,69 @@ public void updateLine(Long id, LineRequest lineRequest) {
 
 	@Transactional
 	public void deleteLine(Long id) {
+		sectionRepository.deleteByLine_Id(id);
```
오잉 갑자기 _id가 튀어나왔네요? _를 넣어야 하나요...?? sections에 line을 line_id로 지정해서 그런가보네요.

직접참조를 했기 때문에 cascade 옵션을쓰면 section을 지우는 작업을 jpa 영속성 컴포넌트가 자동으로 해주기도 해요. :sweat_smile:

hiblue02 commented on src/test/java/subway/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,150 @@
+	@Test
+	void createSectionWithUpStationIsNotEndStationThenFailTest() {
+		// given
+		sectionApiRequest.post(new SectionRequest(4L, 2L, 10), 1L);
```
이미 등록된 역과 노선 ID를 상수로 선언해주면 어떨까요?
그러면 어떤 역/노선을 말하는 것인지 명확하게 알 수 있을 것 같아요.

hiblue02 commented on src/test/java/subway/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,150 @@
+	@DisplayName("지하철 노선의 하행 종점역이 아닌 구간을 제거하면 실패한다.")
+	@DirtiesContext
+	@Test
+	void deleteSectionWithStationIsNotEndThenFailTest() {
```
실패 케이스까지 테스트를 꼼꼼하게 작성해주셨네요! :+1:

hiblue02 commented:
인수테스트를 꼼꼼하게 작성해주셨네요. 특히 테스트 메소드명을 고민해 지은 부분이 인상 깊었어요! 

1. 인수테스트를 한번에 돌리면 깨지는 케이스가 있습니다. 2단계 미션 설명 중 "인수테스트 격리" 부분이 잘 동작하지 않나 봅니다. :sob:
2. 아주 사소한 거지만! 리드미는 프로젝트에 대한 정보를 제공하는 문서인데요. 작업한 내용을 리드미에 정리해보는 건 어떨까요?

hiblue02 reviewed (changes requested):
휘연님! 3단계 미션! 빠르게 잘 해주셨습니다. 👍 
코멘트 남겼으니 확인해주시고, 다시 리뷰요청 주세요!

hhhhhwi committed:
refactor: Line에 Section과의 연관관계 추가
- List<Section> 추가
- addSection, deleteSection : 매개변수 유형 수정
- hasStation : Line이 갖고 있는 구간이 주어진 stationId의 station을 갖고 있는 지 직접 확인하는 메소드 (TODO : 추후 section에서 확인하도록 구현)
- 서비스 레이어 수정 : 한 영속성 안에서 관리되어 추가적으로 section을 save 할 필요 없음. 해당 코드 삭제

hhhhhwi committed:
refactor: 인수테스트 격리

hhhhhwi committed:
refactor: 응답하지 않아도 되는 lineId 삭제

hhhhhwi committed:
refactor: Section을 일급컬렉션으로 구성
- section의 상태와 행위를 한 곳에서 관리하도록 일급컬렉션 Sections 구현
- section과 line 대한 행위 검증 비즈니스 로직을 Entity 에 구현

hhhhhwi committed:
refactor: ExceptionHandler 생성
- IllegalArgumentException에 대해 400 (Bad Request)를 리턴하고, body에 message 파라미터에 에러 메시지를 담아 응답한다.

hhhhhwi committed:
refactor: ExceptionResponse로 응답하도록 수정

hhhhhwi committed:
docs: 리드미 작성

hhhhhwi committed:
refactor: 테스트 코드 리팩토링
- RestApiRequest : path를 각 메소드 별 변수로 받도록 수정
- SectionSteps, LineSteps : 각 인수테스트에 맞는 api 호출을 돕는 Step 클래스 생성. 가독성을 위해 메소드명 한글로 작성

hhhhhwi committed:
refactor: Unused import statement 삭제

hhhhhwi commented:
안녕하세요. 혜영님! 재리뷰요청이 좀 늦었습니다 ㅠ

꼼꼼하게 리뷰해주신 대로 개선을 해봤는데요..!
리뷰해주신 것 처럼 JPA 의 연관관계 개선, 객체의 캡슐화, 테스트 코드 리팩토링에 좀 더 초점을 맞춰서 개선해보았습니다..!

아래 질문과 함께 리뷰 요청 드립니다~!

늘 감사합니다 😄

hhhhhwi commented on src/test/java/subway/LineSteps.java:
```diff
@@ -0,0 +1,32 @@
+import io.restassured.response.Response;
+import subway.dto.LineRequest;
+
+public class LineSteps {
```
테스트 코드를 좀 더 깔끔하고 직관적으로 짜고 싶다는 고민을 할 때 쯤,, 이번주 강의에서 성현님께서 이런 방식으로 구현하시길래 참고하여 구현해보았습니다...!

hiblue02 replied:
:+1:

hhhhhwi commented on src/test/java/subway/LineSteps.java:
```diff
@@ -0,0 +1,32 @@
+	private static final String LINE_API_WITH_ID_URL = "/lines/{id}";
+	private static final RestApiRequest<LineRequest> apiRequest = new RestApiRequest<>();
+
+	public static ExtractableResponse<Response> 노선_생성_요청(String name, String color, Long startStationId, Long endStationId, int distance) {
```
늘 테스트 명이 고민이였는데, 가독성을 위해 한글을 쓰는 것도 하나의 방법..! 이라는 얘기를 듣고 한 번 사용해보았습니다!

hiblue02 replied:
:+1: :+1: 네! @test 를 붙이는 테스트명은 무작위 값을 넣는 곳도 있다고 하네요!

hhhhhwi commented on src/test/java/common/RestApiRequest.java:
```diff
@@ -5,46 +5,54 @@
 	}
 
-	public ExtractableResponse<Response> post(T body) {
+	public ExtractableResponse<Response> post(String path, T body, Object... pathParams) {
```
이전에는 RestApiRequest를 생성하는 시점에 path를 변수로 정의해줬는데, 새로운 유형의 path가 나올 때 마다 새로 선언을 해줘야 하는게 추상화가 깨지는 것 같아 아예 메소드 별로 path 값을 받는 것으로 수정했습니다 😢

hiblue02 replied:
:+1:

hhhhhwi commented on src/main/java/subway/entity/Sections.java:
```diff
@@ -0,0 +1,56 @@
+				.orElseThrow(EntityNotFoundException::new);
+	}
+
+	public List<SectionResponse> convertToSectionResponse() {
```
일급컬렉션을 사용하다보니, dto 변환의 위치가 고민되더라구요 😮 일단 일급컬렉션 안에 구현했으나... 사실 sections의 상태나 행위에 무관한 것 같다는 느낌이 들었습니다..!

hiblue02 replied:
네! 저도 dto변환은 sections의 책임 아니라고 생각해요.  
저는 과제에서,  sections에서 불변객체(collection)을 반환하고, controller나 service 레이어에서 dto로 변환하게 했어요.

hiblue02 commented on README.md:
````diff
@@ -1,2 +1,100 @@
+    Then 등록되지 않고 코드값 500 (Internal sever Error) 을 반환한다.
+    ```
+  - [x] 테스트 메소드 구현
+### 기능 구현
+  - [x] Section 객체 구현
+    - id, lineId, downStationId, upStationId, distance
+  - [x] Line과 Section (1:N)
+  - [x] lineService.createSection
+  - [x] lineService.deleteSection
+  - [x] 예외 처리
+  - [x] 실패 테스트 에러 메시지 검증
+  - [x] 테스트 리팩토링 : 이미 등록된 역과 노선 ID를 상수로 선언
````
리뷰 반영 :100:
구현 사항을 꼼꼼하게 작성해주셨네요!

hiblue02 commented on src/main/java/subway/exception/SubwayExceptionHandler.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+@RestControllerAdvice
+public class SubwayExceptionHandler extends ResponseEntityExceptionHandler {
+	@ExceptionHandler(IllegalArgumentException.class)
+	public ResponseEntity<ExceptionResponse> handleIllegalArgumentException(IllegalArgumentException e) {
+		return ResponseEntity.badRequest().body(new ExceptionResponse(e.getMessage()));
+	}
+
+	@Override
+	protected ResponseEntity<Object> handleHttpRequestMethodNotSupported(HttpRequestMethodNotSupportedException ex, HttpHeaders headers, HttpStatus status, WebRequest request) {
+		return ResponseEntity.internalServerError().body(new ExceptionResponse("예상치 못한 문제가 발생했습니다."));
+	}
+}
```
리뷰 반영 :100:

hiblue02 commented on src/test/java/subway/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,152 @@
+	@Test
+	void createSectionTest() {
+		// when
+		ExtractableResponse<Response> response = SectionSteps.구간_생성_요청(두번째_등록한_종점역, 처음_등록한_종점역, 10, 노선);
```
SectionSteps을 사용하면서 테스트 코드가 훨씬 깔끔해졌네요! :+1:

hiblue02 commented on src/main/java/subway/entity/Sections.java:
```diff
@@ -0,0 +1,56 @@
+		this.sections.remove(section);
+	}
+
+	public boolean hasStation(Long stationId) {
+		for(Section section : sections) {
+			if(stationId.equals(section.getDownStationId())) {
+				return true;
+			}
+
+			if(stationId.equals(section.getUpStationId())) {
+				return true;
+			}
+		}
+
+		return false;
+	}
```
일급컬랙션으로 리팩토링 해주셨네요! 리뷰반영:100:
그리고 역을 찾고, 검증하는 책임을 Stations에 부여해주셨네요. 많이 고민하신 흔적이 보입니다!

hiblue02 commented on src/main/java/subway/entity/Line.java:
```diff
@@ -44,28 +45,73 @@ public String getColor() {
+		this.sections.addSection(section);
+	}
+
+	public void deleteSection(Long stationId) {
+		if(!isEndStation(stationId)) {
+			throw new IllegalArgumentException("노선의 하행 종점역만 제거할 수 있습니다.");
+		}
+
+		Section section = sections.getSectionByDownStationId(stationId);
+
+		if(isStartStation(section.getUpStationId())) {
+			throw new IllegalArgumentException("상행 종점역과 하행 종점역만 있는 노선입니다.");
+		}
+
+		this.endStationId = section.getUpStationId();
+		this.distance = this.distance - section.getDistance();
+
+		sections.removeSection(section);
+	}
```
line에 역할/책임 부여 :+1:

hiblue02 reviewed (approved):
휘연님! 꼼꼼히 리뷰 반영해주셨네요. 
특히 "객체의 책임과 역할"에 대해 고민을 깊게 하신 것이 느껴졌어요. 
이번 단계는 여기서 머지하겠습니다. 😄 
다음 주차 미션도 파이팅입니다! 🔥 🔥
